New Integration: WeatherFlow Cloud Data (Forecast)

jeeftor opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

I wrote `weatherflow` integration which is a local only integration that receives weather data via UDP.  Tempest Weatherflow also provides a [Cloud/REST](https://weatherflow.github.io/Tempest/api/) API. @briis developed a HACS module called [WeatherFlow Forecast](https://github.com/briis/weatherflow_forecast) that utilizes the REST Api. In this PR I am working to bring functionality from that HACS module into core. 

The HACS module provides both Weather Forecast and Cloud based sensor data. I decided to start with the Weather Forecast initially as this should bring the most functionality to HA as we already have a local module to get WeatherFlow data. Once this PR gets approved I'll work to bring in the cloud sensors (and/or) convince @briis to work on the PR himself :)

### Hybrid vs Separate Integrations

I had a discussion with @joostlek about whether I should create a separate integration or extend the existing integration. I started down the path of extending the existing integration but the overall approach to these two integrations is different enough I think it serves to have it as a separate integration. One is entirely `udp` based while this one is entirely `http` based. I could also anticipate some users would want to run both at the same time

<img width="386" alt="image" src="https://github.com/home-assistant/core/assets/6491743/950fff9a-71e1-40aa-abb3-f7099440e8e9">




## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

jeeftor committed:
Initial WeatherFlow Cloud/Forecast commit

joostlek commented:
Lmao, tell me you use PyCharm refactoring without telling me you use PyCharm refactoring

joostlek commented on homeassistant/components/ebox/sensor.py:
```diff
@@ -215,7 +215,7 @@ def __init__(self, username, password, httpsession):
         """Get the latest data from Ebox."""
         try:
-            await self.client.fetch_data()
+            await self.client.fetch_forecast_data()
```
```suggestion
            await self.client.fetch_data()
```

joostlek commented on homeassistant/components/fido/sensor.py:
```diff
@@ -255,7 +255,7 @@ async def async_update(self):
 
         try:
-            await self.client.fetch_data()
+            await self.client.fetch_forecast_data()
```
```suggestion
            await self.client.fetch_data()
```

joostlek commented on homeassistant/components/whirlpool/sensor.py:
```diff
@@ -275,7 +275,7 @@ def available(self) -> bool:
     async def async_update(self) -> None:
         """Update status of Whirlpool."""
-        await self._wd.fetch_data()
+        await self._wd.fetch_forecast_data()
```
```suggestion
        await self._wd.fetch_data()
```

joostlek committed:
Apply suggestions from code review

joostlek commented on homeassistant/components/weatherflow_cloud/__init__.py:
```diff
@@ -0,0 +1,34 @@
+        hass=hass,
+        config_entry=entry,
+    )
+    await data_coordinator._async_update_data()  # pylint: disable=protected-access
```
```suggestion
    await data_coordinator.async_config_entry_first_refresh()
```

joostlek commented on homeassistant/components/weatherflow_cloud/config_flow.py:
```diff
@@ -0,0 +1,108 @@
+        """Validate credentials and obtain station information."""
+
+        session = async_create_clientsession(self.hass)
+        api = WeatherFlow(station_id=station_id, api_token=api_token, session=session)
```
```suggestion
        api = WeatherFlow(station_id, api_token, session)
```
Not sure if they are named arguments, but this is way better imo

joostlek commented on homeassistant/components/weatherflow_cloud/config_flow.py:
```diff
@@ -0,0 +1,108 @@
+        station_info = await api.async_get_station()
+        return station_info
+
+    async def async_step_user(
```
What do you think about flipping the logic? Checkout Suez Water for example

joostlek commented on homeassistant/components/weatherflow_cloud/config_flow.py:
```diff
@@ -0,0 +1,108 @@
+            ),
+            errors=errors or {},
+        )
+
+
+class CannotConnect(HomeAssistantError):
+    """Error to indicate we cannot connect."""
+
+
+class InvalidAuth(HomeAssistantError):
+    """Error to indicate there is invalid auth."""
```
Not used

joostlek commented on homeassistant/components/weatherflow_cloud/config_flow.py:
```diff
@@ -0,0 +1,108 @@
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {vol.Required(CONF_STATION_ID): int, vol.Required(CONF_API_TOKEN): str}
```
Can we fetch a list of all stations?

jeeftor replied:
I'll have to check the docs ... (Mark as TODO)

jeeftor replied:
So it does look like in the docs we have a - get all stations for a user.

<img width="998" alt="image" src="https://github.com/home-assistant/core/assets/6491743/177f4490-d7c7-4ec2-994b-ed3ff2eb9b8b">

So the question is:

- A) User enters credentials - and forecasts get automatically created for all available weather stations... - some what of a redesign - but probably more user friendly at first

- B) User enters credentials - and is then presented a list of possible stations to choose from - they select a station ... if only 1 station its automatically created.

Keeping in mind that the next PR will allow a user to enable Cloud Sensors for a specific weather station - does one option seem better than the other?

If @briis has time to look over this as well - I'd like to get his opinion.

joostlek replied:
I think both would work. I think I would prefer the first. This way is the most simple and I can't really imagine why you don't want to add a certain weather station to your instance
 If you don't want to see it you can disable it.

briis replied:
I have only 1 station, so I am not sure if you get an API Token per Station or per user. If it is per station, then option A is the only possibility. If it is per user, I would go for option B, as the users should have the freedom to select what they want to add per station.

joostlek replied:
It's not what they want to add per station, but what station they want to add

jeeftor replied:
Based on the API I have to assume that 1 toke gives you multiple stations because of this endpoint:

<img width="270" alt="image" src="https://github.com/home-assistant/core/assets/6491743/cc0f3696-1fbd-421c-a341-b7fb6370fd77">

(Happy Boxing Day if any of you are form England?)

joostlek commented on homeassistant/components/weatherflow_cloud/coordinator.py:
```diff
@@ -0,0 +1,113 @@
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(minutes=randrange(25, 35)),
```
que?

jeeftor replied:
I donno seems like maybe we should just do it every 15/30 mins.

@briis why did you select a random 25-35 minute window? 

I'll change it :)

joostlek commented on homeassistant/components/weatherflow_cloud/coordinator.py:
```diff
@@ -0,0 +1,113 @@
+
+
+class WeatherFlowCloudDataUpdateCoordinator(
+    DataUpdateCoordinator["WeatherFLowCloudData"]
```
```suggestion
    DataUpdateCoordinator[WeatherFLowCloudData]
```

joostlek commented on homeassistant/components/weatherflow_cloud/coordinator.py:
```diff
@@ -0,0 +1,113 @@
+    """Unable to connect to the web site."""
+
+
+class WeatherFLowCloudData:
```
Why don't we merge this with the coordinator?

joostlek commented on homeassistant/components/weatherflow_cloud/manifest.json:
```diff
@@ -0,0 +1,13 @@
+  "name": "WeatherflowCloud",
+  "codeowners": ["@briis", "@jeeftor"],
+  "config_flow": true,
+  "dependencies": [],
+  "documentation": "https://www.home-assistant.io/integrations/weatherflow_cloud",
+  "homekit": {},
+  "iot_class": "cloud_polling",
+  "requirements": ["pyweatherflow-forecast==1.0.1"],
+  "ssdp": [],
+  "zeroconf": []
```
Remove empty fields

joostlek commented on homeassistant/components/weatherflow_cloud/strings.json:
```diff
@@ -0,0 +1,26 @@
+        "description": "Set up a WeatherFlow Forecast Station",
+        "title": "WeatherFlow Forecast",
+        "data": {
+          "station_id": "Station Id",
+          "api_token": "Personal Api Token"
```
```suggestion
          "station_id": "Station id",
          "api_token": "Personal api token"
```

joostlek commented on homeassistant/components/weatherflow_cloud/weather.py:
```diff
@@ -0,0 +1,241 @@
+    name: str | None
+    is_metric = hass.config.units is METRIC_SYSTEM
+
+    if (name := config_entry.data.get(CONF_NAME)) and name is None:
+        name = DEFAULT_NAME
+    elif TYPE_CHECKING:
+        assert isinstance(name, str)
```
We don't even store a name

jeeftor replied:
In the current code it is storing a name. Although I went back and forth on this:

<img width="545" alt="image" src="https://github.com/home-assistant/core/assets/6491743/f9010327-8aa5-4a49-a684-084e8bf6c911">

At a minimum we could just store the credentials and at startup figure out which stations to add...

joostlek replied:
Oh lol, I forgot this was a thing. But I actually prefer to store as less as possible. This way if the station name changes HA can pick it up without updating the config entry

jeeftor replied:
I'm hoping we can just store a SINGLE api token which would be awesome.

joostlek commented on homeassistant/components/weatherflow_cloud/weather.py:
```diff
@@ -0,0 +1,241 @@
+        WeatherFlowWeather(coordinator, config_entry.data, False, name, is_metric)
+    ]
+
+    # Add hourly entity to legacy config entries
+    if entity_registry.async_get_entity_id(
+        WEATHER_DOMAIN, DOMAIN, _calculate_unique_id(config_entry.data, True)
+    ):
+        name = f"{name} hourly"
+        entities.append(
+            WeatherFlowWeather(coordinator, config_entry.data, True, name, is_metric)
+        )
```
There's no legacy anymore since this is core

joostlek commented on homeassistant/components/weatherflow_cloud/weather.py:
```diff
@@ -0,0 +1,241 @@
+        self._attr_device_info = DeviceInfo(
+            name="Forecast",
+            entry_type=DeviceEntryType.SERVICE,
+            identifiers={(DOMAIN,)},  # type: ignore[arg-type]
```
Where identifier?

joostlek commented on homeassistant/components/weatherflow_cloud/weather.py:
```diff
@@ -0,0 +1,241 @@
+            manufacturer=MANUFACTURER,
+            model=MODEL,
+            configuration_url=f"https://tempestwx.com/station/{self._config[CONF_STATION_ID]}/grid",
+            hw_version=f"FW V{self._config.get(CONF_FIRMWARE_REVISION, ' - Not Available')}",
```
If its not available, why do we add it?

abmantis commented:
@home-assistant unassign whirlpool

joostlek commented:
Already did :) @abmantis

jeeftor committed:
Update homeassistant/components/weatherflow_cloud/__init__.py

I knew you would know what I was missing!!

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

jeeftor committed:
Update homeassistant/components/weatherflow_cloud/config_flow.py

Thats because you never worked in swift and fell in love with named parameters :)

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

jeeftor committed:
Update homeassistant/components/weatherflow_cloud/strings.json

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

jeeftor committed:
Removed legacy stuff

jeeftor committed:
Big refactor to new backing lib
